### PR TITLE
Add a dedicated ColormapRegistry class

### LINKED ISSUE #18490: Add a method to access the list of registered colormaps
### Describe the issue

I noticed that the linked issues to PR
https://github.com/matplotlib/matplotlib/pull/16991
all want a public way to access the registered colormap names without resorting to importing pyplot. That PR removed the cm.cmap_d dictionary, but left nothing in its place.

There were a few suggestions in this thread: https://github.com/matplotlib/matplotlib/pull/16991#discussion_r415267245

**Summary**

Create a way to access the keys of the colormap registry within the `mpl.cm` module.

### Proposed fix

Some ideas:
`mpl.cm.get_cmap_names()`
`mpl.cm.get_registered_cmaps()`
keeping consistency with pyplot: `mpl.cm.colormaps()`
... something else entirely

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

Following the thoughts in https://github.com/matplotlib/matplotlib/pull/16991#discussion_r415340436, this defines a Mapping subclass `ColorbarRegistry` with the following properties:

- read-only mapping: name -> `Colormap`. - Adding new colormaps is for now still done via `register_cmap()`. The class will grow the ability to add elements later, but that needs more API consideration.
- The returned `Colormaps` are copies, so that the global state is not affected by modification to the returned colormaps. - This is the goal for `get_cmap()` anyway (see its deprecation note).

This is a minimal version, that allows to get the registered colormap names. Closes #18490.

Still to be decided: Where should the instance `colormaps` instance live? In `cm`, in `matplotlib`? Should `cm` be integrated into `colors` anyway?

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by story645** [comment]
I vote leave it in cm for now, but since there's HTML colorbar reprs now (#17888) can you use this to print the list of colormaps?

**Comment by timhoffm** [comment]
You mean as `ColormapRegistry.__repr_html__`? In principle yes. - I'm always a bit cautious on long reprs.

**Comment by story645** [comment]
I hear but like I have that page of the docs printed out cause of how often I look up the colormaps.

**Comment by jklymak** [comment]
guess I'm not particularly following the use-case for this versus just `get_cmap`?  If we just want a list of colormaps, why can't we just store that as a flat list and give the user access to it?  I'm not following why we need a mapping.

**Comment by timhoffm** [comment]
This is motivated in the discussion https://github.com/matplotlib/matplotlib/pull/16991#pullrequestreview-400482771.

The idea is to have a more concise colormap mangement and API. The current state is a bit ad-hoc. You have a private (formerly public) dict and functions `get_cmap`, `register_cmap` working on that. By making the dict private, users lost access to the keys.

You *could* store the names in a flat list, but that would be redundant to the keys of the private dict and needs synchronization. Also, users could manipulate that list - though I'm not too concerned about this.

The whole construct is conceptually a mapping. So why not make it one? Collecting everything into one class makes it easier to manage everything consistently. It makes it also easier to move the user-facing actions out of `cm`. While to be thought about in more detail, I can imagine that the colormaps can be accessed in the future from the top-level module such as: 

~~~
import matplotlib as mpl
mpl.colormaps.register(my_colormap)
cmap = mpl.colormaps['jet']
~~~

**Comment by jklymak** [comment]
OK, so this is just a first step to this class being a new interface for the colormap registry.  Because right now, if I use `cmap_register` this class will no longer return the new colormap, right?  I guess I think at the very least it should return the current state of the registry, not the state at import.  

I'd prefer not having to import `cm` to get this list.  I always import `pyplot`, so I'd be fine with it there, but I understand the aesthetic problem with "we discourage pyplot, except for these 4 things we still think are useful".

**Comment by timhoffm** [comment]
> OK, so this is just a first step to this class being a new interface for the colormap registry.
Correct.

> Because right now, if I use `cmap_register` this class will no longer return the new colormap, right?  I guess I think at the very least it should return the current state of the registry, not the state at import.  

It does. The internal `self._cmaps` is a reference to the same object as `_cmap_registry`:
```
n [9]: cm.colormaps._cmaps is cm._cmap_registry
Out[9]: True
```

> I'd prefer not having to import `cm` to get this list.  I always import `pyplot`, so I'd be fine with it there, but I understand the aesthetic problem with "we discourage pyplot, except for these 4 things we still think are useful". 

Agreed. There should be no need for the user to import `cm`. The nice thing with an object is that we can easily reference it from anywhere. We can instantiate it in `cm` but do `from matplotlib.cm import colormaps` in `matplotlib` and `pyplot` (or we directly instantiate it in `matplotlib` and only import in `pyplot`.

**Comment by greglucas** [comment]
:+1: I like this. I think it makes it clear what is going on and allows for flexibility moving forward.

What would you think about adding a `register()` method directly to this class? It seems like people would want to add/write directly to this registry.
`mpl.colormaps.register(my_new_cmap)`
`mpl.cm.register_cmap()` could just dispatch directly to that method then.

Or even allow people to set the colormaps just like a dict, but raise on name collisions?
`mpl.colormaps['viridis'] = my_new_cmap` -> raise

**Comment by timhoffm** [comment]
@greglucas That's exactly the plan. I'm not completely decided on `register()` vs. dict-setitem, whichnis why it's not yet in the PR. I tend towards an explicit `register()` because we likely won't have a full mutable mapping functionality (or semantic, e.g. concerning overwriting existing keys).

**Comment by timhoffm** [comment]
@greglucas sure. I've added the `register()` method. Open issues:

- Do we want to take over `override_builtins` from `register_cmap()`? In one way or the other, yes. There should be an opt-in way. I'm wondering though if we should treat builtin colormaps differently than others, so maybe `override_existing` would be the better option?
- From above
  > Still to be decided: Where should the instance `colormaps` instance live? In `matplotlib.cm`, or in `matplotlib` itself? Should `cm` be integrated into `colors` anyway?

  In the long run, it should probably be directly below `matplotlib`. For now, we could make it private and  in `matplotlib`:  `from matplotlib.cm import _colormaps as colormap`.

Feedback is welcome, on the open issues as well as any other ideas.

File: lib/matplotlib/cm.py [hunk 1]
```diff
@@ -91,13 +92,66 @@ def _warn_deprecated(self):
...
+        Parameters
+        ----------
+        cmap : matplotlib.colors.Colormap
+           The colormap to register.
+
+        name : str, optional
+           The name for the colormap. If not given, the :attr:`~.Colormap.name`
+           attribute of *cmap* is used.
+        """
+        register_cmap(name, copy.deepcopy(cmap))
```
Comment (greglucas) [inline lib/matplotlib/cm.py:141]: There is a `copy()` method on Colormaps now, so you should be able to call `cmap.copy()` instead and avoid the copy import above.

File: lib/matplotlib/cm.py [hunk 2]
```diff
@@ -91,13 +92,66 @@ def _warn_deprecated(self):
         )
 
 
+class ColormapRegistry(Mapping):
+    r"""
+    Container for colormaps that are known to Matplotlib by name.
+
+    Read access uses a dict-like interface mapping names to `.Colormap`\s.
+    Returned `.Colormap`.s are copies, so that their modification does not
```
Comment (greglucas) [inline lib/matplotlib/cm.py:100]:
```suggestion
    Returned `.Colormap`\s are copies, so that their modification does not
```

**Review by greglucas — COMMENTED** [review]
I agree with `override_existing` being better, but what about changing that to `force=False` instead? The one issue I can see here is that this does mean that import order matters for libraries who `force` over colormaps. But, that seems like bad downstream practice, so should be fine.

I would like to see this Registry at `matplotlib.colormaps`. I don't think matplotlib.cm is imported right away currently, so this would force the import of the Registry now even if people aren't using colormaps in their figures. Probably a minor concern because the lookup table data isn't actually populated immediately.

If you `print(matplotlib.colormaps)` to see what colormaps are available now, would you get the dictionary of key, value pairs, listing the Colormap objects themselves? Right now the `plt.colormaps()` returns a list, so it might be nice to return a list of the keys only.

**Comment by timhoffm** [comment]
I've added a string representation listing all available colormaps which is triggered on `print(matplotlib.colormaps)`.

`plt.colormaps()` is a separate issue. We may or may not want `plt.colormaps` to become the registry as well. That depends a bit on if we want to move the pyplot API also for `get_cmap`. My intuition is not to right now.

`force=False` sounds reasonable. I still want to find out why we choose not to do this in `get_cmap`. Is this just backward compatibility or was there a more fundamental reason.

It's true that `matplotlib` itself does not import `cm`, however pyplot does, so for most practical usages, the colormaps are loaded. A quick test:
```
import matplotlib
%timeit -n 1 -r 1 import matplotlib.cm
```
is about 12.5 ms. That's not the end of the world, but we might want to look into delayed initialization, which also pyplot would benefit from. However, that's an optimization that does not have to be part of this PR.

**Comment by timhoffm** [comment]
Note: I've added `register(..., force=False)`.

It is currently not possible to override builtin colormaps via the `ColormapRegistry` API. That's intentional. I'm questioning the need for such an operation. For now, people can still use `cm.register_cmap(..., override_builtin=True)` if really needed (or overwrite `ColormapRegistry._cmaps`).

**Review by greglucas — COMMENTED** [review]
Overall :+1: I think it is good to have this easily available at the top-level namespace.

It took me a while to find this information in the docs pages. I think it could be added to the main Colormaps docs section. Right now, those all still use `mpl.cm.get_cmap()`.
https://62710-1385122-gh.circle-artifacts.com/0/doc/build/html/tutorials/colors/colormaps.html#sphx-glr-tutorials-colors-colormaps-py

Another area that is missing is how to register a cmap after creating it. The "Creating colormaps in matplotlib" section just shows how to create a colormap, perhaps a final section describing "this is how to register a colormap for future use"
https://62710-1385122-gh.circle-artifacts.com/0/doc/build/html/tutorials/colors/colormap-manipulation.html#creating-colormaps-in-matplotlib

**Comment by timhoffm** [comment]
Docs certainly need adaption. However, I first want to make sure that the API is agreed on before starting to change the docs.
Also technically, we could have this in 3.5 without further doc changes. It's properly introduced, the old way is not deprecated (only discouraged). It would also be ok to change the docs in 3.5.1 if we don't manage that for 3.5.

**Comment by jklymak** [comment]
This seems fine to me.  I feel that we will still have some confusion, as the colormap registry is ephemeral, so why have so much public interface to it.  If I need to create the colormap at startup, I'd just use it directly.  So I think the added docs will be helpful to guide folks and make it clear what is happening here.

File: lib/matplotlib/cm.py [hunk 3]
```diff
@@ -91,13 +91,86 @@ def _warn_deprecated(self):
...
+
+    Returned `.Colormap`\s are copies, so that their modification does not
+    change the global definition of the colormap.
+
+    Additional colormaps can be added via `.ColormapRegistry.register`::
+
+        mpl.colormaps.register(my_colormap)
+    """
+    def __init__(self, cmaps):
+        self._cmaps = cmaps
```
Comment (tacaswell) [inline lib/matplotlib/cm.py:114]: makes sense why to do this as to wrap this around the existing dict, but I am wary of sharing the underlying storage this way.

File: lib/matplotlib/cm.py [hunk 4]
```diff
@@ -91,13 +91,86 @@ def _warn_deprecated(self):
...
+        return iter(self._cmaps)
+
+    def __len__(self):
+        return len(self._cmaps)
+
+    def __str__(self):
+        return ('ColormapRegistry; available colormaps:\n' +
+                ', '.join(f"'{name}'" for name in self))
+
+    def register(self, cmap, *, name=None, force=False):
```
Comment (tacaswell) [inline lib/matplotlib/cm.py:132]: I think we need an deregister as well?  with `force` you can over-write an existing colormap, but I still think we will want a way to remove a key.

I do not think we want to promote this to a MutableMapping and support the full set of del / pop / clear.  The act of adding and removing color maps is a different activity than the act of getting a colormap out of the registry.

**Comment by tacaswell** [comment]
My sense is that long term we want to deprecate `mpl.cm.register_colormap` and friends?  There will have to be some period in the middle where we support both so if that is the plan I'm not worried about the confusion factor.

Getting a bit over my skis, having these in a nice object at least opens a path putting `filter(...)` methods on it that will return a new one with just the color maps of interest (e.g. ask only for perceptually uniform or diverging color maps...assuming we tag them with enough information) which would be super useful for a UI.

**Comment by tacaswell** [comment]
I do not think we are going to get this done in the next few weeks, I propose bumping it to 3.6.

**Comment by greglucas** [comment]
I forget where we stand on deprecating the global colormap mutability and whether this PR is something that needs to get in for that?

This PR mentions that we missed the intermediate state and extended the deprecation period: https://github.com/matplotlib/matplotlib/pull/19766, but I'm not clear if we ever added the intermediate state in 3.5 or not? i.e. is there an action item to add/change the way we get colormaps/warn about it?

xref: https://github.com/matplotlib/matplotlib/issues/19609 for a discussion about `mpl.cm.{cmap_name}`.

File: lib/matplotlib/cm.py (hunk 4 above)
Comment (timhoffm) [reply lib/matplotlib/cm.py:132]: I agree that some sort of deregister should be there at some point. Details would need to be discussed, e.g. if you can deregister built-in cmaps (I think that should not be possible these are reseverd names with fixed meaning and no public API should change them).

Anyway, we currently don't have a deregister function, so there is no hurry with adding that new feature in the initial version of the registry.

File: lib/matplotlib/cm.py (hunk 3 above)
Comment (timhoffm) [reply lib/matplotlib/cm.py:114]: It's very intentional, that the registry shares the underlying storage: For a transition period, the registry and the cmap functions will coexists. Due to API quirks in the functions, it's better if both can directly work on the underlying storage and not one API calling the orther.

If you only have concerns with the constructor accepting a prebuilt mapping, we can change it to creating an empty registry by default and have a private `ColormapRegistry._from_dict()` that we use to inject the existing dict.

**Comment by timhoffm** [comment]
### Overall goal

The registry is a more natural structure for managing colormaps. We have data and corresponding functions that are tightly related. IMHO it's a much nicer API for the user and easier to extend. And we get rid of some API quirks in the functions like `get_cmap(.. lut=...)` or the parameter order `register_cmap(name, cmap)` where name is optional.

### Target state
- the `matplotlib.cm` colormap functions should eventually be removed.
- I'm open for discussion if we keep the pyplot functions or replace that with the registry as well.

### Transition path
While I would like to eventually remove the `matplotlib.cm` functions, I propose a prolonged transition path because there are quite fundamental functions also used in third party libraries:
- For 3.5: introduce the registry as is in this PR; *discourage* the `cm` functions. This is to see if we have missed something important in the design and get user feedback. - Maybe we should declare the registry experimental. 
- For 3.6: Introduce a pending deprecation. This is the time where all downstream libraries should switch to using the registry.
- For 3.7: Deprecate the `cm` functions.
- For 3.9: Remove the `cm` functions.

**I strongly advocate to introduce this PR in 3.5. If everything goes well, we don't need modifications (only extensions) to the registry, which means that downstream libraries would have 3.5 as a minimum dependency when they switch to the registry.**

### On read-only colormaps
The registry colormap access already returns copies of the colormaps and by that implements the change to `get_cmap`, which is scheduled to take effect in 3.6 (#19766). He can be quicker here, because the new API does not have to maintain backward compatibility for the deprecation period.
Other than that, the registry is independent of cmap mutability.

File: doc/users/next_whats_new/colormaps.rst [hunk 5]
```diff
@@ -0,0 +1,18 @@
+Colormaps registry
```
Comment (greglucas) [inline doc/users/next_whats_new/colormaps.rst:1]:
```suggestion
Colormap registry
```

**Review by greglucas — APPROVED** [review]
I agree with @timhoffm's path forward. I think additional capabilities can be added later and doesn't preclude this from going in. I think the chosen location at the base level  `matplotlib.colormaps` with a clear name is a good choice.

One other thought (that I don't think is a great idea because it would add another way to do something) is that you could override `__getattr__` on this Registry and allow access to the built-ins that way so that
`matplotlib.colormaps.viridis` is the same as `matplotlib.colormaps['viridis']`. That may help with moving `matplotlib.cm.viridis` and friends over to this new location.

**Comment by jklymak** [comment]
My objection is a little more fundamental;  I'm not sure why we need a registry at all, so rather than hoisting it to the top level I would tend to discourage passing colormaps by string.

**Comment by tacaswell** [comment]
The ship on not supporting colormaps-by-string sailed long ago and I think deprecating that in any meaningful way is out of the question.  If we are going to support it, I would rather support it with a nice object which can be both more discover-able and inspect-able than the current state.

**Comment by timhoffm** [comment]
> One other thought (that I don't think is a great idea because it would add another way to do something) is that you could override `__getattr__` on this Registry and allow access to the built-ins that way so that
`matplotlib.colormaps.viridis` is the same as `matplotlib.colormaps['viridis']`. That may help with moving matplotlib.cm.viridis and friends over to this new location.

Thanks for the proposal. It’s always good to add ideas to the solution space. However, I agree that this is not a great idea 😀. As you mentioned, this is just another way to achieve the same thing. And worse: it dilutes the mapping concept. The registry is a simple mapping: name -> Colormap, which is canonically associated with item access. The attributes would pollute the register namespace and introduce an unnecessary asymmetry between built in and third party color maps. I think that matplotlib.cm.viridis was an error in the first place. I don't want to replicate this to the registry. OTOH the future of these cm functions is a separate discussion and possibly targeted at 3.6.

File: lib/matplotlib/__init__.py [hunk 6]
```diff
@@ -1445,3 +1445,8 @@ def inner(ax, *args, data=None, **kwargs):
 _log.debug('interactive is %s', is_interactive())
 _log.debug('platform is %s', sys.platform)
 _log.debug('loaded modules: %s', list(sys.modules))
+
+
+# workaround: we must defer colormaps import to after loading rcParams, because
+# colormap creation depends on rcParams
+from matplotlib.cm import _colormaps as colormaps
```
Comment (tacaswell) [inline lib/matplotlib/__init__.py:1452]: Could we handle this in the module level `__getattr__` now that we have than in place?

File: lib/matplotlib/cm.py (hunk 3 above)
Comment (tacaswell) [reply lib/matplotlib/cm.py:114]: I would be happy warning in the init docstring saying we keep a reference to the input dict.

It makes total sense why you are doing it this way, just set off alarm bells in my head (we have had some cases at BNL where we did not _mean_ to do this and ended up with hard-to-debug action-at-a-distance bugs!).

File: lib/matplotlib/cm.py (hunk 4 above)
Comment (tacaswell) [reply lib/matplotlib/cm.py:132]: We have `unregister_cmap` which went out in 3.4 via https://github.com/matplotlib/matplotlib/pull/15127

https://github.com/matplotlib/matplotlib/blob/1761822ada3f8a081872bce419c54f4b4237fab2/lib/matplotlib/cm.py#L197-L199

**Review by tacaswell — APPROVED** [review]
I leave to @timhoffm 's discretion if we want to add the unregister method in this pass so 👍🏻 self-merging if you want to defer it.

**Comment by jklymak** [comment]
Let's merge.  Someone can always add an unregister later.

**Pull request merged**